### DISCUSSION TITLE
use jscad.app renderer in vanilla app?

### DISCUSSION THREAD

**Discussion by tsdexter** [hrgdavor/jscadui#68] [General]
Thanks for the great work on this @hrgdavor  (and others). Is there an example of how to use the three renderer, with the dark mode theme, no requirement for the full server, importing/editing scripts, etc... in a vanilla html/js page where we can just pass our own `main()` function to render or even just the return value of main (array of model objects)?

**Comment by hrgdavor** [comment]
https://github.com/hrgdavor/jscadui/tree/main/apps/model-page is something like that.
in main.js line for theme is: `const theme = themes.light`

This is  a quickly made prototype. It has been a while since it is updated, so it is currently broken. this whole repo is currently in prototyping mode, with few things that are relatively stable.

IT is advisable to use the worker to exec the script, to not block the main thread. Also feel free to jump in the discord channel.

**Comment by danmarshall** [comment]
+1 on the great work! Are there any updates on how to use this?

**Reply by hrgdavor** [reply]
it is getting better slowly, and it is still in progress and needs refactoring.

https://jscad.app is a demo that could be some kind of how-to  ... by looking at the source for making sth on your own ATM that is the best I got.

it also works if not at the root, which is nice for placing as part of a website, and not being the main thing. I use this feature to publish previews of work in progress that is not yet published like this one https://github.com/hrgdavor/jscadui/pull/95

One PR that is important is: https://github.com/hrgdavor/jscadui/pull/93 . it is consolidating  the use-case of working with project folders that are drag dropped from file system. 

Also for standalone apps, I need to separate into module all this complex dnd folder project stuff out of the demo app.
Not all standalone apps will want to be like some IDE thingy. And without that dnd part, worker can still run complex projects that have imports reachable by fetch ... for example from github or a website that has proper CORS, or complex projects  at the same origin.

**Reply by hrgdavor** [reply]
@danmarshall if you have something specific in mind, I am bit more motivated to smash up some code when something concrete is needed ... otherwise I slowly do some bits, as time permits. Feel free to reach out on the discord channel.

**Reply by tsdexter** [reply]
@danmarshall if you just want a threejs renderer (vs. IDE type environment) that you can pass an array of models to (or a main func), I found it not too complicated to wire up a svelte component that renders them with threlte/threejs (and a bit of code borrowed from here). I am also working on extracting it out into a standalone component but competing priorities are slowing that down right now

**Reply by danmarshall** [reply]
Thanks @hrgdavor & @tsdexter ! (I'm also short on cycles for extracurricular activities 😆). 
My 3 projects have various needs:
* jscad-now has a bad memory leak. Somewhere in my animation loop to regl-renderer.
* jscad-gallery is outdated
* Maker.js is outdated, uses very old OpenJscad code. Would be nice to add a viewer to the playground.
Also thanks for mentioning the Discord @hrgdavor - very interesting in there! I found out about Manifold which is also very interesting!

**Reply by hrgdavor** [reply]
@danmarshall np, it is nice to see u are still around, whatever little time you can scrap :) ..
Yes manifold is very interesting, and I hope to integrate it to jscad ... either officially, or at least as sth that can be combined with jscad.

stick around, it may take some time, but I will let you know when I extract and simplify usefull stuff out od jscad.app demo. Some things in jscadui have taken nice shape to be reusable, but I have published to npm only 3mf format and gizmo.
